source/brailleDisplayDrivers/braillePen12.py

```diff
@@ -0,0 +1,182 @@
+#brailleDisplayDrivers/braillePen12.py
+#A part of NonVisual Desktop Access (NVDA)
+#This file is covered by the GNU General Public License.
+#See the file COPYING for more details.
+#Copyright (C) 2014 Harpo Sp. z o. o. <http://www.harpo.com.pl>
+
+import braille
+import queueHandler
+from logHandler import log
+from ctypes import *
+from ctypes.wintypes import *
+import time
+import config
+import inputCore
+import hwPortUtils
+import brailleInput
+import wx
+
+READ_INTERVAL = 50
+
+BP12_KEYCALLBACKHANDLE = WINFUNCTYPE(None,c_ubyte,c_ubyte,c_ubyte,c_ubyte,c_ubyte,c_ubyte,c_ubyte,c_ubyte)
+
+BP12_BRAILLE_KEYS = ("dot1", "dot2", "dot3", "dot4", "dot5", "dot6", "dot7", "dot8")
+BP12_CONTROL_KEYS=("leftScroll","joyAction","joyLeft","joyUp","joyDown","joyRight","space", "rightScroll")
+
+#Try to load BP12TI32.dll
+try:
+	BP12Lib=windll[r"brailleDisplayDrivers\BP12TI.dll"]
+except:
+	BP12Lib=None
+
+def _findPorts():
+	for portInfo in hwPortUtils.listComPorts(onlyAvailable=True):
+		try:
+			btName = portInfo["bluetoothName"]
+		except KeyError:
+			continue
+		if btName.startswith("EL12-"):
+			yield portInfo["port"]
+
+class BrailleDisplayDriver(braille.BrailleDisplayDriver):
+	name = "braillePen12"
+	# Translators: The name of a series of braille displays.
+	description = _("BP12 and EL12 series")
+	isThreadSafe = True
+	
+	@classmethod
+	def check(cls):
+		return bool(BP12Lib)
+
+	def __init__(self):
+		super(BrailleDisplayDriver,self).__init__()
+		self.numCells = 12	
+		for port in _findPorts():
+			if not port:
+				raise RuntimeError("No BP12 display found")
+				continue
+			portName=u"{port}".format(port=port)
+			self._keyCallbackInst = BP12_KEYCALLBACKHANDLE(self._keyCallback)
+			result=0
+			try:
+				result=BP12Lib.open(portName,self._keyCallbackInst)
+			except:
+				result=0
+			if result >0:
+				self._connectedState=True
+				break
+		else:
+			raise RuntimeError("No BP12 display found")
+
+	def terminate(self):
+		super(BrailleDisplayDriver, self).terminate()
+		try:
+			BP12Lib.close()
+		except:
+			pass
+
+	def handleResponses(self):
+		pass
+		
+	def display(self, cells):
+		try:
+			arr="".join([chr(x) for x in cells])
+			if self._connectedState==True:
+				result=BP12Lib.sendBraille(0,len(cells),arr)
+				if result<0:
+					self._connectedState=False
+		except:
+			pass
```

Comment:
As above, rather than trying to gracefully handle disconnects, raise an exception so NVDA knows about it. Note that #1555 fixes things so NVDA will disable the display isconnect if an exception gets raised here. In future, #1271 will mean that the display can then be detected again when it becomes available.